docs/design.md

```diff
@@ -0,0 +1,59 @@
+* and also a command-line tool named `sparkctl` for working with the operator. 
+
+The following diagram shows how different components interact and work together.
+
+![Architecture Diagram](architecture-diagram.png)
+
+Specifically, a user uses the `sparkctl` (or `kubectl`) to create a `SparkApplication` object. The `SparkApplication` controller receives the object through a watcher from the API server, creates a submission carrying the `spark-submit` arguments, and sends the submission to the *submission runner*. The submission runner submits the application to run and creates the driver pod of the application. Upon starting, the driver pod creates the executor pods. While the application is running, the *Spark pod monitor* watches the pods of the application and sends status updates of the pods back to the controller, which then updates the status of the application accordingly. 
+
+### Controller, Submission Runner, and Spark Pod Monitor
+
+The `SparkApplication` controller, or CRD controller in short, watches events of creation, updates, and deletion of `SparkApplication` objects in any namespaces in a Kubernetes cluster, and acts on the watch events. When it receives a new `SparkApplication` object, it prepares a submission and sends the submission to the submission runner, which actually submits the application to run in the Kubernetes cluster. The submission includes the list of arguments for the `spark-submit` command. The submission runner has a configurable number of workers for submitting applications to run in the cluster. 
+
+The controller is also responsible for updating the status of a `SparkApplication` object with the help of the Spark pod monitor, which watches Spark pods and update the `SparkApplicationStatus` field of corresponding `SparkApplication` objects based on the status of the pods. The Spark pod monitor watches events of creation, updates, and deletion of Spark pods, creates status update messages based on the status of the pods, and sends the messages to the controller to process. The controller maintains the collection of `SparkApplication` objects of running applications and update the status of them based on the messages received.
+
+As part of preparing a submission for a newly created `SparkApplication` object, the controller parses the object and adds configuration options for adding certain annotations to the driver and executor pods of the application. The annotations are later used by the Spark pod initializer to configure the pods before they start to run. For example,if a Spark application needs a certain Kubernetes ConfigMap to be mounted into the driver and executor pods, the controller adds an annotation that specifies the name of the ConfigMap to mount. Later the Spark pod initializer sees the annotation on the pods and mount the ConfigMap to the pods.
```

Comment:
Is there a reason this has to be done as an initializer (or mutating admission), rather than just propagating the information directly down the chain? For example, I would have expected that the SparkApplication controller could create a SparkDriver object, which contains the information necessary for the SparkDriver controller to create executor Pods that have all the needed ConfigMaps mounted to begin with.

Do you need to inject things indirectly because you don't have control over the driver code that's creating the executor Pods? If so, why?

Reply:
Because the operator has no control over the driver/executor creation code in Spark itself. The operator simply calls the `spark-submit` command, which in turns calls the submission client that creates the driver pod. The Kubernetes mode of Spark supports a limited set of customization points for pods.